src/coreclr/jit/valuenum.cpp

```diff
@@ -15034,3 +15061,82 @@ void ValueNumStore::PeelOffsets(ValueNum* vn, target_ssize_t* offset)
         }
     }
 }
+
+//--------------------------------------------------------------------------------
+// VNWalkPhis: Walk the given PhiDef and call a callback on each Phi argument (its VN).
+//
+// Arguments:
+//    phiDefFunc   - VNF_PhiDef function
+//    walkPhiVnsFn - The callback function to call on each Phi argument.
+//    maxDepth     - Maximum depth to handle recursive PhiDef chains.
+//
+// Return Value:
+//    true if the walk was successful, false if the walk was aborted.
+//
+bool ValueNumStore::VNWalkPhis(VNFuncApp phiDefFunc, WalkPhiVnsFn walkPhiVnsFn, int maxDepth)
+{
+    assert(maxDepth > 0);
+    assert(phiDefFunc.m_func == VNF_PhiDef);
+
+    // VNF_PhiDef(LclNum, SsaNum, VNF_Phi)
+    unsigned lclNum     = phiDefFunc.m_args[0];
+    ValueNum nextPhiArg = phiDefFunc.m_args[2];
+
+    if ((lclNum == static_cast<unsigned>(-1)) || (nextPhiArg == NoVN))
+    {
+        return false;
+    }
+
+    // Now we iterate over Phi arguments, e.g.
+    //  2 args: VNF_Phi(SSA1, SSA2)
+    //  3 args: VNF_Phi(SSA1, VNF_Phi(SSA2, SSA3))
+    //  4 args: VNF_Phi(SSA1, VNF_Phi(SSA2, VNF_Phi(SSA3, SSA4)))
+    // etc.
+    //
+    while (nextPhiArg != NoVN)
+    {
+        ValueNum  current = nextPhiArg;
+        VNFuncApp phiArgFuncApp;
+        if (GetVNFunc(nextPhiArg, &phiArgFuncApp) && phiArgFuncApp.m_func == VNF_Phi)
+        {
+            current    = phiArgFuncApp.m_args[0];
+            nextPhiArg = phiArgFuncApp.m_args[1];
+        }
+        else
+        {
+            // Ok this was the last Phi arg
+            nextPhiArg = NoVN;
+        }
+
+        // Extract the real VN from the Phi arg
+        unsigned phiArgSsaNum = ConstantValue<unsigned>(current);
+        ValueNum phiArgVN     = m_pComp->lvaTable[lclNum].GetPerSsaData(phiArgSsaNum)->m_vnPair.Get(VNK_Conservative);
+
+        // Now check the actual VN arg, if it's yet another PhiDef, walk it recursively
+        VNFuncApp argFuncApp;
+        if (GetVNFunc(phiArgVN, &argFuncApp) && (argFuncApp.m_func == VNF_PhiDef))
+        {
+            if (maxDepth == 1)
+            {
+                JITDUMP("optWalkPhiVNs: maxDepth reached - bail out.\n");
+                return false;
```

Comment:
It's unfortunate that this is indistinguishable from the callback returning false. I think it would be better if the function returned a 3-way enum instead of `Aborted, Completed, HitLimit`.

More generally, in many cases there is no need to conservatively give up in the cyclic cases -- e.g. `IsVNNeverNegative` can work just fine when it hits a cyclic phi if all the other args are non-negative. So I wonder if we could instead design the helper to be more powerful -- for example, call it `VNsInSSAWebSatisfy(predicate)` -- that checks if all the VNs of the SSA web satisfy some predicate. It requires keeping hash set around for the VNs we have already visited, but in my experience that is not that expensive (e.g. SCEV analysis does that).

Reply:
If we generalize it enough (`VisitSSAWeb`) then we could even rewrite `VNForMapSelectWork` to be based on it...